File: native/dart/packages/blocks_runtime/test/version_test.dart
import 'package:blocks_runtime/src/user_agent.dart';
import 'package:test/test.dart';

/// Validates the generated version constant and user-agent token grammar.
void main() {
  group('blocksUserAgentToken', () {
    // The token format this library commits to: `aws-blocks-<lang>/<semver>` —
    // a bounded lowercase-alphanumeric language segment plus a strict semver.
    final grammar = RegExp(
      r'^aws-blocks-([a-z][a-z0-9]{0,15})/(\d+\.\d+\.\d+(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?)$',
    );

    test('matches the expected token format', () {
      expect(
        grammar.hasMatch(blocksUserAgentToken),
        isTrue,
        reason:
            '$blocksUserAgentToken does not match the expected '
            'aws-blocks-<lang>/<semver> format',
      );
    });

    // Rows whose user agent does not contain `aws-blocks` do not appear in
    // reporting, so renaming the token would lose this library's traffic.
    test('keeps the aws-blocks prefix the warehouse filters on', () {
      expect(blocksUserAgentToken.contains('aws-blocks'), isTrue);
    });

    test('embeds the current version', () {
      expect(
        blocksUserAgentToken,
        equals('aws-blocks-dart/$blocksRuntimeVersion'),
      );
    });
  });

  group('blocksRuntimeVersion', () {
    test('is valid semver', () {
      final semver = RegExp(r'^\d+\.\d+\.\d+(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$');
      expect(
        semver.hasMatch(blocksRuntimeVersion),
        isTrue,
        reason: '$blocksRuntimeVersion is not valid semver',
      );
    });

    test('is not a placeholder', () {
      expect(blocksRuntimeVersion, isNot(equals('0.0.0')));
    });
  });
}

File: native/kotlin/runtime/src/commonTest/kotlin/com/aws/blocks/kotlin/VersionTest.kt
package com.aws.blocks.kotlin

import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldMatch
import kotlin.test.Test

class VersionTest {

    // Token format: aws-blocks-<lang>/<version>. The language is 1-16 lowercase alphanumeric
    // characters starting with a letter; version is numeric X.Y.Z with an optional prerelease and no build metadata.
    private val grammar = Regex("""^aws-blocks-([a-z][a-z0-9]{0,15})/(\d+\.\d+\.\d+(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?)$""")

    @Test
    fun `user agent token matches the expected format`() {
        blocksUserAgentToken shouldMatch grammar
    }

    @Test
    fun `user agent token keeps the aws-blocks prefix`() {
        blocksUserAgentToken shouldContain "aws-blocks"
    }

    @Test
    fun `user agent token embeds the runtime version`() {
        blocksUserAgentToken shouldBe "aws-blocks-kotlin/$blocksRuntimeVersion"
    }

    @Test
    fun `runtime version is valid semver`() {
        blocksRuntimeVersion shouldMatch Regex("""^\d+\.\d+\.\d+(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$""")
    }

    @Test
    fun `runtime version is not a placeholder`() {
        blocksRuntimeVersion shouldNotBe "0.0.0"
    }
}
